README.rst
==========
cryptopals
==========

.. image:: https://img.shields.io/travis/svenpruefer/cryptopals.svg
:target: https://travis-ci.org/svenpruefer/cryptopals

Solutions to the cryptopals exercises in Scala by Sven Prüfer.

* Free software: GNU General Public License v3
* Documentation: https://svenpruefer.github.io/cryptopals/

Impressum
---------

You can find the authors website on Musmehl_.

.. _Musmehl: https://sven.musmehl.de


HISTORY.rst
=======
History
=======

0.1.0 (2018-03-02)
------------------

Work started on first problem sheet


src/test/scala/de/musmehl/cryptopals/Set1/problem1.scala
package de.musmehl.cryptopals.Set1

import org.scalatest.FunSuite

class CryptopalsProblems extends FunSuite {

    test("Challenge 1") {
        val input = HexString("49276d206b696c6c696e6720796f757220627261696e206c696b65206120706f69736f6e6f7573206d757368726f6f6d")
        val output = Base64String("SSdtIGtpbGxpbmcgeW91ciBicmFpbiBsaWtlIGEgcG9pc29ub3VzIG11c2hyb29t")

        assert(input.toBase64String == output)
    }

    test("Challenge 2") {
        val input1 = HexString("1c0111001f010100061a024b53535009181c")
        val input2 = HexString("686974207468652062756c6c277320657965")
        val output = HexString("746865206b696420646f6e277420706c6179")

        assert(input1.xor(input2) == output)
    }
}


src/test/scala/de/musmehl/cryptopals/Set1/CryptopalsProblems.scala
package de.musmehl.cryptopals.Set1

import org.scalatest.FunSuite

class CryptopalsProblems extends FunSuite {

    test("Challenge 1") {
        val input = HexString("49276d206b696c6c696e6720796f757220627261696e206c696b65206120706f69736f6e6f7573206d757368726f6f6d")
        val output = Base64String("SSdtIGtpbGxpbmcgeW91ciBicmFpbiBsaWtlIGEgcG9pc29ub3VzIG11c2hyb29t")

        assert(input.toBase64String == output)
    }

    test("Challenge 2") {
        val input1 = HexString("1c0111001f010100061a024b53535009181c")
        val input2 = HexString("686974207468652062756c6c277320657965")
        val output = HexString("746865206b696420646f6e277420706c6179")

        assert(input1.xor(input2) == output)
    }

    test("Challenge 3") {
        val input = HexString("1b37373331363f78151b7f2b783431333d78397828372d363c78373e783a393b3736")
        val (minimum, result) = decodeXorLetterEncryption(input)

        println(s"Found key: $minimum with resulting string: ${result.toAsciiString}")
    }
}


src/main/scala/de/musmehl/cryptopals/Set1/HexString.scala
package de.musmehl.cryptopals.Set1

import de.musmehl.cryptopals.Set1

/**
  * Class of a hex string supplying various methods for dealing with hex encoded strings
  *
  * @param stringContent the hex string
  */
case class HexString(stringContent: String) {

    /**
      * xor bitwise with that
      *
      * @param that the second argument as a HexString
      * @return the bitwise xor'd HexString
      */
    def xor(that: HexString): HexString = {
        val xoredSequence = this.toByteArray.zip(that.toByteArray).map(x => x._1 ^ x._2).map(_.toByte)
        HexString(xoredSequence.map(byteToHexMap).mkString)
    }

    /**
      * yields the hex string as a sequence of bytes according to the hexToByteMap.
      *
      * Notice that every hex character is an actual byte, i.e. 8 bits instead of the necessary 4 bits.
      */
    val toByteArray: Seq[Byte] = stringContent.map(Set1.hexToByteMap)

    /**
      * converts a hex string to a base64 encoded string
      *
      * @return the hex string as a base64 string
      */
    def toBase64String: Base64String = {
        val sequenceOfHexBytes: Seq[Seq[Byte]] = toByteArray.grouped(6).toList.map(_.padTo[Byte, Seq[Byte]](6, 0))
        val sequenceOfBase64Bytes: Seq[Seq[Byte]] = sequenceOfHexBytes.map(mapTripleOfBytesToQuadrupleOfBase64Bytes)
        Base64String(sequenceOfBase64Bytes.flatten.map(byteToBase64Map).mkString)
    }

    /**
      * takes a sequence of six (!) bytes corresponding to six hex characters, concatenates their bits and splits this
      * into four bytes corresponding to four base64 characters
      *
      * @param array the six hex characters as bytes
      * @return the quadruple of base64 characters as bytes
      * @example suppose array is given by the hex string '3f0a11', this corresponds to bytes 0x03, 0x0f, 0x00, 0x0a,
      *          0x01 and 0x01 which in turn is in bits 0000 0011, 0000 1111, 0000 0000, 0000 1011, 0000 0001
      *          and 0000 0001. Concatenating the corresponding last four bits each we get the 24 bits
      *          001111110000101100010001 splitting into 001111, 110000, 101100 and 010001, yielding the integers 15,
      *          48, 34 and 17, corresponding to bytes 0x0f, 0x30, 0x2c and 0x11 or base64 characters P, w, i and R
      */
    private def mapTripleOfBytesToQuadrupleOfBase64Bytes(array: Seq[Byte]): Seq[Byte] = {
        require(array.length == 6, "Only works for sixtets of bytes")
        require(array.forall(_ < 16), "Every byte needs to correspond to a hex character, i.e. needs to be less than 16")

        val piece1 = ((array.head << 2) + (array(1) >>> 2)).toByte
        val piece2 = (((array(1) % 4) << 4) + array(2)).toByte
        val piece3 = ((array(3) << 2) + (array(4) >>> 2)).toByte
        val piece4 = (((array(4) % 4) << 4) + array(5)).toByte

        Seq(piece1, piece2, piece3, piece4)
    }

    private def mapPairOfHexBytesToByte(array: Seq[Byte]): Byte = {
        require(array.length == 2, "Only works for pairs of bytes")
        require(array.forall(_ < 16), "Every byte needs to correspond to a hex character, i.e. needs to be less than 16")

        ((array.head << 4) + array(1)).toByte
    }

    val toAsciiString: String = {
       toByteArray.grouped(2).toList.map(_.padTo[Byte, Seq[Byte]](2,0)).map(mapPairOfHexBytesToByte).map(_.toChar).mkString
    }
}

object HexString {
    def fromAsciiString(string: String): HexString = {
        require(string.forall(_.toByte < 128), "String needs to consist of Ascii characters only")

        HexString(string.map(_.toByte).flatMap(x => List(byteToHexMap((x >>> 4).toByte), byteToHexMap((x % 16).toByte))).mkString)
    }

}


src/main/scala/de/musmehl/cryptopals/Set1/package.scala
package de.musmehl.cryptopals

package object Set1 {
    val base64ToByteMap: Map[Char,Byte] = Map(
        ('A',0), ('Q',16), ('g',32), ('w',48),
        ('B',1), ('R',17), ('h',33), ('x',49),
        ('C',2), ('S',18), ('i',34), ('y',50),
        ('D',3), ('T',19), ('j',35), ('z',51),
        ('E',4), ('U',20), ('k',36), ('0',52),
        ('F',5), ('V',21), ('l',37), ('1',53),
        ('G',6), ('W',22), ('m',38), ('2',54),
        ('H',7), ('X',23), ('n',39), ('3',55),
        ('I',8), ('Y',24), ('o',40), ('4',56),
        ('J',9), ('Z',25), ('p',41), ('5',57),
        ('K',10), ('a',26), ('q',42), ('6',58),
        ('L',11), ('b',27), ('r',43), ('7',59),
        ('M',12), ('c',28), ('s',44), ('8',60),
        ('N',13), ('d',29), ('t',45), ('9',61),
        ('O',14), ('e',30), ('u',46), ('+',62),
        ('P',15), ('f',31), ('v',47), ('/',63)
    )

    val byteToBase64Map: Map[Byte,Char] = Map(
        (0, 'A'), (16, 'Q'), (32, 'g'), (48, 'w'),
        (1, 'B'), (17, 'R'), (33, 'h'), (49, 'x'),
        (2, 'C'), (18, 'S'), (34, 'i'), (50, 'y'),
        (3, 'D'), (19, 'T'), (35, 'j'), (51, 'z'),
        (4, 'E'), (20, 'U'), (36, 'k'), (52, '0'),
        (5, 'F'), (21, 'V'), (37, 'l'), (53, '1'),
        (6, 'G'), (22, 'W'), (38, 'm'), (54, '2'),
        (7, 'H'), (23, 'X'), (39, 'n'), (55, '3'),
        (8, 'I'), (24, 'Y'), (40, 'o'), (56, '4'),
        (9, 'J'), (25, 'Z'), (41, 'p'), (57, '5'),
        (10, 'K'), (26, 'a'), (42, 'q'), (58, '6'),
        (11, 'L'), (27, 'b'), (43, 'r'), (59, '7'),
        (12, 'M'), (28, 'c'), (44, 's'), (60, '8'),
        (13, 'N'), (29, 'd'), (45, 't'), (61, '9'),
        (14, 'O'), (30, 'e'), (46, 'u'), (62, '+'),
        (15, 'P'), (31, 'f'), (47, 'v'), (63, '/')
    )

    val hexToByteMap: Map[Char,Byte] = Map(
        ('0', 0),
        ('1', 1),
        ('2', 2),
        ('3', 3),
        ('4', 4),
        ('5', 5),
        ('6', 6),
        ('7', 7),
        ('8', 8),
        ('9', 9),
        ('a', 10),
        ('b', 11),
        ('c', 12),
        ('d', 13),
        ('e', 14),
        ('f', 15)
    )

    val byteToHexMap: Map[Byte, Char] = Map(
        (0, '0'),
        (1, '1'),
        (2, '2'),
        (3, '3'),
        (4, '4'),
        (5, '5'),
        (6, '6'),
        (7, '7'),
        (8, '8'),
        (9, '9'),
        (10, 'a'),
        (11, 'b'),
        (12, 'c'),
        (13, 'd'),
        (14, 'e'),
        (15, 'f')
    )

    def scoreEnglishPlainTextString(string: String): Double = {
        val relativeFrequency = Map[Char, Double](
            ('a', 0.08167),('b', 0.01492),
            ('c', 0.02782),('d', 0.04253),
            ('e', 0.12702),('f', 0.02228),
            ('g', 0.02015),('h', 0.06094),
            ('i', 0.06966),('j', 0.00153),
            ('k', 0.00772),('l', 0.04025),
            ('m', 0.02406),('n', 0.06749),
            ('o', 0.07507),('p', 0.01929),
            ('q', 0.00095),('r', 0.05987),
            ('s', 0.06327),('t', 0.09056),
            ('u', 0.02758),('v', 0.00978),
            ('w', 0.02360),('x', 0.00150),
            ('y', 0.01974),('z', 0.00074)
        )

        val totalNumberOfLetters = string.count(_.isLetter)
        val frequencyInString =
            (for (letter <- relativeFrequency.keys)
                yield (letter, string.count(_.toLower == letter).toDouble / totalNumberOfLetters)).toMap[Char, Double]

        relativeFrequency.keys.map(x => math.pow(relativeFrequency(x) - frequencyInString(x), 2)).sum
        + math.pow((totalNumberOfLetters - string.length).toDouble / totalNumberOfLetters, 2)
    }

    def decodeXorLetterEncryption(hexString: HexString): (Char, HexString) = {
        require(hexString.stringContent.length % 2 == 0, "HexString should contain an even number of hex characters")

        val n = hexString.stringContent.length / 2

        val deviations = (for (letter <- "abcdefghijklmnopqrstuvwxyz") yield
            (letter, scoreEnglishPlainTextString(
                HexString.fromAsciiString(List.fill(n)(letter).mkString).xor(hexString).toAsciiString)))
            .toMap[Char, Double]
        val minimum = deviations.find(x => deviations(x._1) == deviations.values.min).get._1
        (minimum, HexString.fromAsciiString(List.fill(n)(minimum).mkString).xor(hexString))
    }


}
